src/callbacks/check_gradients.cpp

```diff
@@ -116,11 +228,11 @@ void check_gradients::do_check_gradients(model& m) const {
   // For simplicity, we assume f(chi) ~ f(x), and | f'''''(xi) | ~ 1.
   // If step size is not specified, then we choose h so that
   //   E_fl <= sqrt(epsilon)
-  const DataType epsilon = std::pow(std::numeric_limits<DataType>::epsilon(), 0.9);
-  const DataType step_size = (m_step_size > DataType{0} ?
+  const EvalType epsilon = std::pow(std::numeric_limits<EvalType>::epsilon(), 0.9);
```

Comment:
This means that we expect the numerical accuracy corresponding to `double`, so I'd expect all of our gradient checking with `float` layers to fail. Now that I think about it, having a default error tolerance is tricky with mixed precision models. I think we should make the default correspond to `float`, and require the user to manually put in a tolerance if they're working with a model with non-`float` layers.


```suggestion
  const EvalType epsilon = std::pow(std::numeric_limits<float>::epsilon(), 0.9);
```